components/server/src/workspace/workspace-starter.ts

```diff
@@ -663,7 +667,9 @@ export class WorkspaceStarter {
                 log.info({ instanceId: instance.id }, "starting instance");
                 return (await manager.startWorkspace(ctx, startRequest)).toObject();
             } catch (err: any) {
-                if ("code" in err && err.code !== grpc.status.OK && lastInstallation !== "") {
+                if ("code" in err && err.code === grpc.status.RESOURCE_EXHAUSTED) {
+                    throw err;
```

Comment:
I'm not sure here: So far, this would have maybe been handled by the re-try. Now we always fail-fast. :thinking: 

But on the other hand this is a clear indication to not try again. :+1: 

@andrew-farries Did you explore whether we get additional info from that middleware, e.g. when to try next?

Reply:
> I'm not sure here: So far, this would have maybe been handled by the re-try. Now we always fail-fast. 🤔
>
> But on the other hand this is a clear indication to not try again. 👍

The way errors are 'handled' in `tryStartWorkspace` is odd; any gRPC errors are silently ignored and the next available cluster is tried instead. AIUI in SaaS there will only ever be one cluster on which a workspace can be started. If there is no cluster on which a workspace can be started then the method returns `undefined`. This means that gRPC errors get swallowed and we only get a generic `instanceStartFailure` in the metrics.

By failing early in the case of a resource exhausted error we at least get to see that as a separate failure reason in the metrics, but we don't get the retry behaviour.

> Did you explore whether we get additional info from that middleware, e.g. when to try next?

AIUI the rate limiting middleware: 

https://github.com/gitpod-io/gitpod/blob/6548573f86a4b30e96a94caf5b54184c9627faac/components/common-go/grpc/ratelimit.go#L159-L208

will block on forwarding to the actual RPC until allowed to do so by the rate limit. It only returns `RESOURCE_EXHAUSTED` in error cases or if the rate limiter is configured to do so by setting `block = false`. It doesn't return early with an indication of when to try again

Reply:
> It doesn't return early with an indication of when to try again.

Thx, understood. Something to improve for later I guess (/cc @kylos101 )

> If there is no cluster on which a workspace can be started then the method returns undefined

Ah, I thought it would re-throw the last error again :grimacing: Maybe it would make sense to adjust that. But not in this PR. :lotus_position: 

> The way errors are 'handled' in tryStartWorkspace is odd; any gRPC errors are silently ignored and the next available cluster is tried instead. AIUI in SaaS there will only ever be one cluster on which a workspace can be started. 

That's true, this is in a half-builtout state. Originally the idea was to switch to different clusters (even across regions!) whenever there is a (probably cluster-local) error on `startWorkspace`. I think we never moved forward because we feared a degradation of service due to db-sync :see_no_evil: . But this would indeed increase resilience I guess.